iree/compiler/Dialect/Flow/Transforms/MatmulToMMT4dConversion.cpp

```diff
@@ -0,0 +1,242 @@
+    }
+
+    // This is for float only matmul for now. Integer data type might require
+    // r.h.s layout change.
+    if (!lhsType.getElementType().isF32() || !rhsType.getElementType().isF32())
+      return failure();
+
+    int m = lhsType.getShape()[0];
+    int n = rhsType.getShape()[1];
+    int k = rhsType.getShape()[0];
+
+    if (m % M0Size != 0 || n % N0Size != 0 || k % K0Size != 0) return failure();
+
+    int m1 = m / M0Size;
+    int n1 = n / N0Size;
+    int k1 = k / K0Size;
+
+    auto lhs4D = expandTo4D(loc, rewriter, lhs, {m1, M0Size, k1, K0Size});
+    auto rhs4D = expandTo4D(loc, rewriter, rhs, {k1, K0Size, n1, N0Size});
+    auto dst4D = expandTo4D(loc, rewriter, dst, {m1, M0Size, n1, N0Size});
+
+    auto lhs4DT = transposeOperand(loc, rewriter, lhs4D, {0, 2, 1, 3});
+    auto rhs4DT = transposeOperand(loc, rewriter, rhs4D, {2, 0, 1, 3});
+    auto dst4DT = transposeOperand(loc, rewriter, dst4D, {0, 2, 1, 3});
+
+    auto mmt4DResult = rewriter.create<linalg::Mmt4DOp>(
+        loc, dst4DT.getType(), ValueRange{lhs4DT, rhs4DT}, ValueRange{dst4DT});
+
+    auto mmt4dResultTransposed =
+        transposeOperand(loc, rewriter, mmt4DResult.getResult(0), {0, 2, 1, 3});
+
+    Value result = collapseTo2D(loc, rewriter, mmt4dResultTransposed, {m, n});
+
+    rewriter.replaceOp(matmulOp, ArrayRef<Value>{result});
+
+    return success();
+  }
+
+ private:
+  int M0Size;
+  int N0Size;
+  int K0Size;
+};
+
+/// Canonicalizes [linalg.init_tensor -> linalg.fill -> linalg.generic] ->
+/// [linalg.init_tensor -> linalg.fill] where linalg.generic does only copy e.g
+/// a transpose.
+struct FoldFillGenericOpPattern : public OpRewritePattern<linalg::GenericOp> {
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    if (genericOp.getNumInputs() != 1) return failure();
+    if (genericOp.getNumOutputs() != 1) return failure();
+
+    // Check linalg.generic does have copy only semantics.
+    if (genericOp.getNumParallelLoops() != genericOp.getNumLoops())
+      return failure();
+    auto results =
+        llvm::to_vector<4>(genericOp.getBody()->getOps<linalg::YieldOp>());
+    if (results.size() != 1) return failure();
+    if (results[0].values().size() != 1) return failure();
+    auto blockArgument = results[0].values()[0].dyn_cast<BlockArgument>();
+    if (!blockArgument || blockArgument.getArgNumber() != 0) return failure();
+
+    auto input = genericOp.inputs()[0];
+
+    auto outputType =
+        genericOp.outputs()[0].getType().dyn_cast<RankedTensorType>();
+
+    // TODO: To enable dynamic shapes we need to apply the same permutation on
+    // init tensor sizes.
+    if (!outputType || !outputType.hasStaticShape()) return failure();
+
+    auto fillOp = dyn_cast<linalg::FillOp>(input.getDefiningOp());
+    if (!fillOp) return failure();
+
+    auto initTensorOp =
```

Comment:
Unnecessary. Dropping lines 184 - 187 here would still give you what you want. You dont need to check for the `linalg.init_tensor` here.

Reply:
Done but, I think we need because we we can have `... -> subtensor -> fill -> linalg_generic` which in this case this canonicalization isn't correct

Reply:
Why is it not correct? All those operations have copy semantics.

Reply:
Replacing `subtensor -> fill -> linalg_generic` with `init_tensor-> fill` isn't correct

Reply:
I have no idea why you are saying that is wrong.

First of all `subtensor -> fill` is a weird pattern. That should never exist. THe `init_tensor` that the `fill` takes is just for its shape`. It does not matter that this is a `subtensor`. I think you are confusing buffer semantics and tensor semantics.

Reply:
Isn't correct not the right phrasing I would say (not complete) or it losses the information. I agree with you this is a tensor copy semantics but its useful to not drop this information in the IR. This isn't a concern for this PR and this pattern will be upstreamd can be re-visit.